cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -834,6 +1000,65 @@ private static MDSKey appendDefaultVersion(String recordType, Type typeOfT, MDSK
     }
   }
 
+  /**
+   * Returns a ProgramId given the MDS key
+   *
+   * @param key the MDS key to be used
+   * @return ProgramId created from the MDS key
+   */
+  private static ProgramId getProgramID(MDSKey key) {
+    MDSKey.Splitter splitter = key.split();
+
+    // Format : recordType, ns, app, <version>, type, program, <ts>, runid
+    // <version> -> might or might not be present based on whether the upgrade thread has changed the record
+    // <ts> -> will be present if the record type is runRecordComplete
+
+    String recordType = splitter.getString();
+    String namespace = splitter.getString();
+    String application = splitter.getString();
+    String appVersion = ApplicationId.DEFAULT_VERSION;
+    String type;
+    String program;
+
+    List<String> splits = new ArrayList<>();
+    while (true) {
```

Comment:
No, we cannot use hasRemaining() since we don't know whether the next part is integer or string.